File: NQueen.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                board[i][j] = '.';
            }
        }
        dfs(res, board, n, 0);
        return res;
    }
    
    private List<String> boardToRes(char[][] board) {
        List<String> res = new ArrayList<>();
        for (char[] row : board) {
            res.add(String.valueOf(row));
        }
        return res;
    }
    
    private void dfs(List<List<String>> res, char[][] board, int n, int r) {
        if (r >= n) {
            res.add(boardToRes(board));
            return;
        }
        for (int c = 0; c < n; ++c) {
            if (isValid(board, r, c, n)) {
                board[r][c] = 'Q';
                dfs(res, board, n, r + 1);
                board[r][c] = '.';
            }
        }
    }
    
    private boolean isValid(char[][] board, int r, int c, int n) {
        for (int i = 0; i != r; ++i) {
            if (board[i][c] == 'Q') {
                return false;
            }
        }
        for (int i = r - 1, j = c - 1; i >= 0 && j >= 0; --i, --j) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        for (int i = r - 1, j = c + 1; i >= 0 && j < n; --i, ++j) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        return true;
    }
}

File: N-Queen.cpp
class Solution {
  public:
    vector < vector < string >> solveNQueens(int n) {
      vector < int > rowPos(n, -1);
      vector < int > colPos(n, -1);
      vector < vector < string >> boards;

      backtracking(n, 0, rowPos, colPos, boards);
      return boards;
    }

  void backtracking(int n, int from, vector < int > & rowPos, vector < int > & colPos,
    vector < vector < string >> & boards) {
    if (from == n) {
      vector < string > board(n, string(n, '.'));
      for (int row = 0; row < n; row++) {
        int col = rowPos[row];
        board[row][col] = 'Q';
      }
      boards.push_back(board);
      return;
    }
    for (int col = 0; col < n; col++) {
      bool valid = isValid(n, from, col, rowPos, colPos);
      if (valid) {
        rowPos[from] = col;
        colPos[col] = from;
        backtracking(n, from + 1, rowPos, colPos, boards);
        rowPos[from] = -1;
        colPos[col] = -1;
      }
    }
  }

  bool isValid(int n, int row, int col, vector < int > & rowPos, vector < int > & colPos) {
    if (rowPos[row] != -1)
      return false;
    if (colPos[col] != -1)
      return false;
    int tmpRow = row;
    int tmpCol = col;
    while (true) {
      tmpRow++;
      tmpCol++;

      if (tmpRow == n || tmpCol == n) break;
      if (rowPos[tmpRow] == tmpCol) return false;
    }
    tmpRow = row;
    tmpCol = col;
    while (true) {
      tmpRow--;
      tmpCol--;

      if (tmpRow < 0 || tmpCol < 0) break;
      if (rowPos[tmpRow] == tmpCol) return false;
    }
    tmpRow = row;
    tmpCol = col;
    while (true) {
      tmpRow++;
      tmpCol--;

      if (tmpRow == n || tmpCol < 0) break;
      if (rowPos[tmpRow] == tmpCol) return false;
    }
    tmpRow = row;
    tmpCol = col;
    while (true) {
      tmpRow--;
      tmpCol++;

      if (tmpRow < 0 || tmpCol == n) break;
      if (rowPos[tmpRow] == tmpCol) return false;
    }
    return true;
  }
};
//
